src/subprocess-posix.cc

```diff
@@ -315,15 +387,20 @@ bool SubprocessSet::DoWork() {
   }
 
   interrupted_ = 0;
-  int ret = pselect(nfds, &set, 0, 0, 0, &old_mask_);
+  s_sigchld_received = 0;
+  int ret = pselect(nfds, (nfds > 0 ? &set : nullptr), 0, 0, 0, &old_mask_);
   if (ret == -1) {
     if (errno != EINTR) {
       perror("ninja: pselect");
       return false;
     }
+    CheckConsoleProcessTerminated();
```

Comment:
Thanks a lot, this is definitely cleaner!

However, shouldn't this function also be called in the non-error case as well? E.g. if there is a lot of file descriptor activity + a console process termination at the same time (similar to the situation explained in the comment on lines 401-403), pselect() would not necessarily return an error, and the `s_sigchld_received` flag change would be missed.  Same goes for ppoll() below. What do you think?

Reply:
You're right! I'm changing that.